src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,143 @@
+            size++;
+        }
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException(index);
+        }
+        if (size == 0 || index == size) {
+            add(value);
+        } else {
+            Node<T> newElement = new Node<>();
+            newElement.value = value;
+            newElement.next = getElement(index);
+            getElement(index).prev = newElement;
 
+            if (index > 0) {
+                getElement(index - 1).next = newElement;
+                newElement.prev = getElement(index - 1);
+            } else {
+                first = newElement;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add((T) list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getElement(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldItem = getElement(index).value;
+        getElement(index).value = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T removedItemValue = getElement(index).value;
+
+        if (size == 1) {
+            first = null;
+            lastElement = null;
+        } else {
+            if (getElement(index).next == null) {
+                lastElement = getElement(index - 1);
+            } else {
+                getElement(index + 1).prev = getElement(index).prev;
+            }
+            if (getElement(index).prev == null) {
+                first = getElement(index + 1);
+            } else {
+                getElement(index - 1).next = getElement(index).next;
+            }
+        }
+        size--;
+        return removedItemValue;
     }
 
     @Override
     public boolean remove(T t) {
+        int counter = 0;
+        while (counter < size) {
+            if (getElement(counter).value == t
+                    || (t != null && t.equals(getElement(counter).value))) {
+                remove(counter);
+                return true;
+            }
+            counter++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getElement(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(index);
+        }
+        Node<T> element = first;
+        int counter = 0;
+        while (counter < index) {
+            element = element.next;
+            counter++;
+        }
+        return element;
+    }
+
+    private class Node<T> {
```

Comment:
```suggestion
    private static class Node<T> {
```
we can use static class here